File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/RestrictDeserializedTypes.LosFormatter.cs
using System;
using System.Web.UI;

namespace SonarAnalyzer.UnitTest.TestCases
{
    internal class RestrictDeserializedTypes
    {
        public void DefaultConstructor()
        {
            new LosFormatter(); // Noncompliant - MAC filtering should be enabled
        }

        public void LiteralExpression()
        {
            new LosFormatter(false, ""); // Noncompliant - MAC filtering should be enabled
            new LosFormatter(false, new byte[0]); // Noncompliant - MAC filtering should be enabled

            new LosFormatter(true, ""); // Compliant - MAC filtering is enabled
            new LosFormatter(true, new byte[0]); // Compliant - MAC filtering is enabled

            new LosFormatter(macKeyModifier: new byte[0], enableMac: false); // Noncompliant
            new LosFormatter(macKeyModifier: new byte[0], enableMac: true); // Compliant
        }

        public void FunctionParameter(bool condition)
        {
            new LosFormatter(condition, ""); // Compliant - unsure about condition value

            if (condition)
            {
                new LosFormatter(condition, ""); // Compliant - MAC filtering is enabled
            }
            else
            {
                new LosFormatter(condition, ""); // Noncompliant - MAC filtering should be enabled
            }
        }

        public void LocalVariables()
        {
            var trueVar = true;
            new LosFormatter(trueVar, ""); // Compliant - MAC filtering is enabled

            var falseVar = false;
            new LosFormatter(falseVar, ""); // Noncompliant - MAC filtering should be enabled
        }

        public void TernaryOp(bool condition)
        {
            var falseVar = condition ? false : false;
            new LosFormatter(falseVar, ""); // Noncompliant - MAC filtering should be enabled

            var trueVar = condition ? true : true;
            new LosFormatter(trueVar, "");

            new LosFormatter(condition ? false : true, "");
        }

        public LosFormatter ExpressionBodyFalse() =>
            new LosFormatter(false, ""); // Noncompliant - MAC filtering should be enabled

        public LosFormatter ExpressionBodyTrue() =>
            new LosFormatter(true, ""); // Compliant - MAC filtering is enabled

        public void InLambdaFunction()
        {
            Func<LosFormatter> createSafe = () => new LosFormatter(true, "");  // Compliant - MAC filtering is enabled
            Func<LosFormatter> createUnsafe = () => new LosFormatter(true, "");  // Compliant - FP: lambda functions are not scanned by symbolic execution
        }

        public LosFormatter SwitchExpression(bool condition)
        {
            return condition switch
            {
                true => new LosFormatter(condition, ""),
                false => new LosFormatter(condition, "") // Compliant - FP: lambda functions are not scanned by symbolic execution
            };
        }

        public void DataFlow()
        {
            var condition = false;
            condition = true;
            if (condition)
            {
                new LosFormatter(condition, "");
            }
            else
            {
                new LosFormatter(condition, ""); // Unreachable
            }

            condition = false;
            if (condition)
            {
                new LosFormatter(condition, ""); // Unreachable
            }
            else
            {
                new LosFormatter(condition, ""); // Noncompliant - MAC filtering should be enabled
            }

            new LosFormatter(new bool(), "");
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/RestrictDeserializedTypes.LosFormatter.vb
Imports System
Imports System.Web.UI

Class RestrictDeserializedTypes
    Public Sub DefaultConstructor()
        Dim formatter = New LosFormatter()                                                      ' Noncompliant {{Serialized data signature (MAC) should be verified.}}
    End Sub

    Public Sub LiteralExpression()
        Dim formatter1 = New LosFormatter(False, "")                                            ' Noncompliant
        Dim formatter2 = New LosFormatter(False, New Byte(-1) {})                               ' Noncompliant
        Dim formatter3 = New LosFormatter(True, "")                                             ' Compliant - MAC filtering is enabled
        Dim formatter4 = New LosFormatter(True, New Byte(-1) {})                                ' Compliant
        Dim formatter5 = New LosFormatter(macKeyModifier:=New Byte(-1) {}, enableMac:=False)    ' Noncompliant
        Dim formatter6 = New LosFormatter(macKeyModifier:=New Byte(-1) {}, enableMac:=True)     ' Compliant
    End Sub

    Public Sub FunctionParameter(condition As Boolean)
        Dim formatter1 = New LosFormatter(condition, "")                                        ' Noncompliant

        If condition Then
            Dim formatter2 = New LosFormatter(condition, "")                                    ' Compliant
        Else
            Dim formatter3 = New LosFormatter(condition, "")                                    ' Noncompliant
        End If
    End Sub

    Public Sub LocalVariables()
        Dim trueVar = True
        Dim formatter1 = New LosFormatter(trueVar, "")                                          ' Compliant
        Dim falseVar = False
        Dim formatter2 = New LosFormatter(falseVar, "")                                         ' Noncompliant
    End Sub

    Public Sub TernaryOp(condition As Boolean)
        Dim falseVar = If(condition, False, False)
        Dim formatter1 = New LosFormatter(falseVar, "")                                         ' Noncompliant
        Dim trueVar = If(condition, True, True)
        Dim formatter2 = New LosFormatter(trueVar, "")                                          ' Compliant
        Dim formatter3 = New LosFormatter(If(condition, False, True), "")                       ' Noncompliant
    End Sub

    Public Function UnsafeReturnValue() As LosFormatter
        Return New LosFormatter(False, "")                                                      ' Noncompliant
    End Function

    Public Function SafeReturnValue() As LosFormatter
        Return New LosFormatter(True, "")
    End Function

    Public Sub InLambdaFunction()
        Dim createSafe As Func(Of LosFormatter) = Function() New LosFormatter(True, "")         ' Compliant
        Dim createUnsafe As Func(Of LosFormatter) = Function() New LosFormatter(False, "")      ' Noncompliant
    End Sub

    Public Function Switch(condition As Boolean) As LosFormatter
        Select Case condition
            Case True
                Return New LosFormatter(condition, "")                                          ' Compliant
            Case Else
                Return New LosFormatter(condition, "")                                          ' Noncompliant
        End Select
    End Function

    Public Sub DataFlow()
        Dim condition = False
        condition = True

        If condition Then
            Dim formatter1 = New LosFormatter(condition, "")
        Else
            Dim formatter2 = New LosFormatter(condition, "")                                    ' Unreachable
        End If

        condition = False

        If condition Then
            Dim formatter3 = New LosFormatter(condition, "")                                    ' Unreachable
        Else
            Dim formatter4 = New LosFormatter(condition, "")                                    ' Noncompliant
        End If

        Dim formatter5 = New LosFormatter(New Boolean(), "")                                    ' Noncompliant
        Dim formatter6 = New LosFormatter(Nothing, "")                                          ' Noncompliant
    End Sub

End Class
